### ISSUE TITLE
hardlinks over ssh

### ISSUE THREAD

**Issue by dudman** [trapexit/mergerfs#347]
when I attempt to backup with backintime-qt4 to a remove server over ssh to my mergerfs /backup i'm told that the filesystem dosn't support hard-links.

On the computer with mergerfs I can obviously create hardlinks, also on a remote machine with this /backup mount over sshfs I can also happily create hardlinks.

Its just when using backintime-qt4, probably an issue with backintime-qt4, but wanted to ask here, incase.

Great tool

**Comment by trapexit** [comment]
I can't speak to backintime-qt4 or sshfs. Please read the docs with regard to link.

https://github.com/trapexit/mergerfs#rename--link

**Comment by dudman** [comment]
Thanks, I've read it a few times, and its a little beyond me although very well written... I understand thats its basically handled, hard links that is ? Am I right ? I've just tested it and the link count goes up, and works. But I guess the inode no or underlying (something) isn't equal when backintime-qt checks... heres the code which does the check and fails on backintime-qt
`
inodes = []
        for tmp in (remote_tmp_dir_1, remote_tmp_dir_2):
            for line in output_split:
                m = re.match(r'^(\d+).*?%s' %tmp, line)
                if m:
                    inodes.append(m.group(1))

        logger.debug('remote inodes: ' + ' | '.join(inodes), self)
        if len(inodes) == 2 and inodes[0] != inodes[1]:
            raise MountException(_('Remote host %s doesn\'t support hardlinks') % self.host)

`

**Comment by trapexit** [comment]
Seems like a strange way to test if links work. Unless I'm missing something it should simply try to link a file and if it succeeds take the filesystem's word that it did. You'll need to talk with the author. I don't know why they are doing it that way.

There may be ways for me to work around this but it's possibly complicated.

**Comment by trapexit** [comment]
Might be a simpler fix than expected if I understand what they are doing correctly. Mind testing for me?

**Comment by trapexit** [comment]
Check out the "[inode](https://github.com/trapexit/mergerfs/tree/inode)" branch and use `use_ino` option when using mergerfs. I'm not 100% the code won't cause issues. I've fired off an email to the FUSE mailinglist to find out more about the feature I'm using.

**Comment by dudman** [comment]
happy to test anything... 

I did leave an issue also at the backintime-qt git, but no response yet...

**Comment by trapexit** [comment]
Let me know if the inode branch addresses the issue. If I get back from the FUSE team that what I'm doing is OK I'll include the change in the next release.

**Comment by dudman** [comment]
I checked out the inode address, built it, and copied the mergerfs binary over the system one (backup of course)... remounted the share... and still get the same message about fs not supporting hard links... 

mergerfs -v gives
-bash: merger: command not found
[root@arch ~]# mergerfs -v
mergerfs version: 2.17.0-4-g9a261a7
FUSE library version: 2.9.7
fusermount version: 2.9.7
using FUSE kernel interface version 7.19

just to check I'm running the right version

**Comment by trapexit** [comment]
That doesn't appear to be correct:

```
$ git clone -b inode http://github.com/trapexit/mergerfs
$ make
...
$ ./mergerfs -v
mergerfs version: 2.17.0-5-g75db423
$ sudo ./mergerfs -f -oallow_other,defaults,use_ino <srcmounts> <dest>
```

If you run `stat` on a file you should see a rather large number (9345989279745) vs a pretty small number (258).

**Comment by dudman** [comment]
My mistake I never added the option use_ino... thanks it works, problem gone... I'm interested in what the problem was ? Thanks

Should I continue using that version for a while, is it safe ?

**Comment by trapexit** [comment]
From my testing it should be fine but I'm waiting to hear back from the FUSE mailinglist to see if what I'm doing is valid.

It's somewhat low level. libfuse's API automatically generates inodes. I was just using that feature. It has no way of knowing two files are the same and as such gives two files with the same inode different ones. What `use_ino` does is allow me to tell it what values to use instead and I'm passing through a modified version of the original which means links will show up correctly as having the same inode.

It's a little tricky because while it's not required inodes normally should be unique (unless it's a hard link). Since I'm combining multiple drives there is obviously the problem that the same inode value could exist on multiple drives.

The inode value is 64bit but many filesystems only use 32bits. Similarly the device number associated with a device (man 2 stat, look for st_dev)  is 64 bit but often is only 16bits of data. So I create an inode which is the original inode value logically or'ed with (st_dev << 32). Which should generally give unique values in a way that's pretty legit in terms of expectations from other programs.

If I find out what I've done is OK at the fuse layer then I'll probably make it the default behavior.

**Comment by dudman** [comment]
Great... thanks for the explanation, which I understand, gives me confidence that I have a little idea of whats going on under the hood.

**Comment by Germar** [comment]
> Seems like a strange way to test if links work. Unless I'm missing something it should simply try to link a file and if it succeeds take the filesystem's word that it did.

This is to cover more point of failures than just the filesystem. I use `rsync --link-dest` to create hard-links and so does the test try to create hard-links with `rsync --link-dest`, too.

**Comment by trapexit** [comment]
Addressed by #349
